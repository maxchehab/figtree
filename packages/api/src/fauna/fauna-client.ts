import { Client, query as q, ExprVal } from 'faunadb';

import { FaunaEntity } from './interfaces/fauna-entity.interface';
import { LoginRequest } from '../common/interfaces/login-request.interface';
import { User } from '../common/interfaces/user.interface';

export class FaunaClient {
  private readonly client = new Client({
    secret: process.env.FAUNA_SECRET as string,
  });

  async userByLoginRequest(code: string): Promise<FaunaEntity<User> | null> {
    try {
      return this.client.query(
        q.Get(q.Match(q.Index('users_by_login_request_code'), code)),
      );
    } finally {
      return null;
    }
  }

  async updateUser(ref: ExprVal, data: Partial<User>) {
    this.client.query(q.Update(ref, { data }));
  }

  async userByToken(token: string): Promise<FaunaEntity<User> | null> {
    try {
      return this.client.query(
        q.Get(q.Match(q.Index('users_by_token'), token)),
      );
    } finally {
      return null;
    }
  }

  async createLoginRequest(
    code: string,
    createdAt: Date,
  ): Promise<FaunaEntity<LoginRequest>> {
    return this.client.query(
      q.Create(q.Collection('login_requests'), {
        data: { code, createdAt: createdAt.toISOString() },
      }),
    );
  }

  async loginRequestByCode(
    code: string,
  ): Promise<FaunaEntity<LoginRequest> | null> {
    try {
      return this.client.query(
        q.Get(q.Match(q.Index('login_requests_by_code'), code)),
      );
    } finally {
      return null;
    }
  }
}
