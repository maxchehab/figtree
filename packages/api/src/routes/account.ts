import { AuthStatus } from '../common/interfaces/auth-status.enum';
import { FaunaClient } from '../fauna/fauna-client';
import { HttpException } from '../common/exceptions/http.exception';
import { Lambda } from '../common/util/lambda.util';
import getBearerToken from '../common/util/get-bearer-token.util';

const faunaClient = new FaunaClient();

export default Lambda(async (req, res) => {
  const token = getBearerToken(req);
  const user = await faunaClient.userByToken(token);

  if (!user) {
    throw new HttpException(400, 'Invalid token.');
  }

  const { data } = user;

  if (data.auth_status !== AuthStatus.LoggedIn) {
    throw new HttpException(400, 'Invalid token.');
  }

  return res.status(200).json({ user: data });
});
