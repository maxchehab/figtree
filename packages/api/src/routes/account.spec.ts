import axios from 'axios';
import http from 'http';
import listen from 'test-listen';

import { AuthStatus } from '../common/interfaces/auth-status.enum';
import { FaunaClient } from '../fauna/fauna-client';
import account from './account';
import closeServer from '../test/close-server';
import createTestLambda from '../test/create-test-lambda';

describe('account', () => {
  let server: http.Server;
  let url: string;

  beforeAll(async () => {
    server = createTestLambda(account);
    url = await listen(server);
  });

  afterAll(async () => {
    await closeServer(server);
  });

  describe('with a user found by the token provided', () => {
    describe(`with the user's 'auth_status' as Logged In`, () => {
      beforeEach(() => {
        jest
          .spyOn(FaunaClient.prototype, 'userByToken')
          .mockImplementationOnce(async () => ({
            ref: '@ref:123',
            data: {
              email: 'example@gmail.com',
              auth_status: AuthStatus.LoggedIn,
              id: 'usr_123',
              login_request_code: 'code_123',
              token: 'token_123',
            },
          }));
      });

      it('returns 200, with the user', async () => {
        const { status, data } = await axios.get(url, {
          validateStatus: () => true,
          headers: { authorization: 'Bearer token_123' },
        });

        expect(status).toEqual(200);
        expect(data).toMatchSnapshot();
      });
    });

    describe(`without the user's 'auth_status' as Logged In`, () => {
      beforeEach(() => {
        jest
          .spyOn(FaunaClient.prototype, 'userByToken')
          .mockImplementationOnce(async () => ({
            ref: '@ref:123',
            data: {
              email: 'example@gmail.com',
              auth_status: AuthStatus.LoggedOut,
              id: 'usr_123',
              login_request_code: 'code_123',
              token: 'token_123',
            },
          }));
      });

      it('returns 400, Invalid token.', async () => {
        const { status, data } = await axios.get(url, {
          validateStatus: () => true,
          headers: { authorization: 'Bearer token_123' },
        });

        expect(status).toEqual(400);
        expect(data).toEqual({
          message: 'Invalid token.',
        });
      });
    });
  });

  describe('without a user found by the token provided', () => {
    beforeEach(() => {
      jest
        .spyOn(FaunaClient.prototype, 'userByToken')
        .mockImplementationOnce(async () => null);
    });

    it('returns 400, Invalid token.', async () => {
      const { status, data } = await axios.get(url, {
        validateStatus: () => true,
        headers: { authorization: 'Bearer token_123' },
      });

      expect(status).toEqual(400);
      expect(data).toEqual({
        message: 'Invalid token.',
      });
    });
  });
});
