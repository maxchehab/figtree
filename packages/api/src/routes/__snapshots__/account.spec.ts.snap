// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account with a user found by the token provided with the user's 'auth_status' as Logged In returns 200, with the user 1`] = `
Object {
  "user": Object {
    "auth_status": "LOGGED_IN",
    "email": "example@gmail.com",
    "id": "usr_123",
    "login_request_code": "code_123",
    "token": "token_123",
  },
}
`;
